pandas/core/arrays/categorical.py

```diff
@@ -0,0 +1,2331 @@
+        For explicit `dtype`, the `inferred_categories` are cast to the
+        appropriate type.
+
+        Parameters
+        ----------
+
+        inferred_categories : Index
+        inferred_codes : Index
+        dtype : CategoricalDtype or 'category'
+
+        Returns
+        -------
+        Categorical
+        """
+        from pandas import Index, to_numeric, to_datetime, to_timedelta
+
+        cats = Index(inferred_categories)
+
+        known_categories = (isinstance(dtype, CategoricalDtype) and
+                            dtype.categories is not None)
+
+        if known_categories:
+            # Convert to a specialzed type with `dtype` if specified
+            if dtype.categories.is_numeric():
+                cats = to_numeric(inferred_categories, errors='coerce')
+            elif is_datetime64_dtype(dtype.categories):
+                cats = to_datetime(inferred_categories, errors='coerce')
+            elif is_timedelta64_dtype(dtype.categories):
+                cats = to_timedelta(inferred_categories, errors='coerce')
+
+        if known_categories:
+            # recode from observation oder to dtype.categories order
+            categories = dtype.categories
+            codes = _recode_for_categories(inferred_codes, cats, categories)
+        elif not cats.is_monotonic_increasing:
+            # sort categories and recode for unknown categories
+            unsorted = cats.copy()
+            categories = cats.sort_values()
+            codes = _recode_for_categories(inferred_codes, unsorted,
+                                           categories)
+            dtype = CategoricalDtype(categories, ordered=False)
+        else:
+            dtype = CategoricalDtype(cats, ordered=False)
+            codes = inferred_codes
+
+        return cls(codes, dtype=dtype, fastpath=True)
+
+    @classmethod
+    def from_codes(cls, codes, categories, ordered=False):
+        """
+        Make a Categorical type from codes and categories arrays.
+
+        This constructor is useful if you already have codes and categories and
+        so do not need the (computation intensive) factorization step, which is
+        usually done on the constructor.
+
+        If your data does not follow this convention, please use the normal
+        constructor.
+
+        Parameters
+        ----------
+        codes : array-like, integers
+            An integer array, where each integer points to a category in
+            categories or -1 for NaN
+        categories : index-like
+            The categories for the categorical. Items need to be unique.
+        ordered : boolean, (default False)
+            Whether or not this categorical is treated as a ordered
+            categorical. If not given, the resulting categorical will be
+            unordered.
+        """
+        try:
+            codes = np.asarray(codes, np.int64)
+        except (ValueError, TypeError):
```

Comment:
Note: I changed this from `except Exception` for the linter.

Reply:
+1